File: frontend/src/lib/api/staff.ts
// 직원 인건비 상세 API — 고용형태·보험·주휴수당까지 반영한 '실제 나가는 돈'
// (백엔드 /api/v1/staff. 이름·시급 등록 자체는 operation의 /employees가 담당)
import { apiFetch } from './client';

const auth = (token: string) => ({ Authorization: `Bearer ${token}` });

export type EmploymentType = 'part_time' | 'part_time_15' | 'full_time' | 'manager';
export type InsuranceType = 'four' | 'two' | 'none';

export type StaffProfile = {
  employee_id: number;
  employment_type: EmploymentType;
  pay_type: 'hourly' | 'monthly';
  monthly_salary: number;
  weekly_hours: number;
  insurance: InsuranceType;
  weekly_holiday_pay: boolean;
  hired_on: string | null;
  memo: string | null;
  unset?: boolean; // 아직 상세를 채우지 않은 직원
  work_days?: string[];
  color?: string;
};

export type LaborCost = {
  weekly_hours: number;
  monthly_hours: number;
  /** 근무시간 근거 — schedule: 근무 달력 / profile: 직접 입력한 소정근로시간 / none: 아직 모름 */
  hours_source: 'schedule' | 'profile' | 'none';
  pay_type: 'hourly' | 'monthly';
  hourly_rate: number;
  effective_hourly: number;
  base_pay: number;
  weekly_holiday_pay: number;
  gross_pay: number;
  withholding_tax: number;
  net_pay: number;
  owner_insurance: Record<string, number>;
  owner_burden: number;
  total_cost: number;
  below_min_wage: boolean;
  min_wage: number;
  disclaimer: string;
};

/** 근무 가능 시간 한 칸 — day_of_week는 0=월 … 6=일 (백엔드 기피시간과 같은 규칙) */
export type AvailabilityWindow = {
  day_of_week: number;
  start_hour: number;
  end_hour: number;
};

export type StaffMember = {
  id: number;
  name: string;
  role: string;
  hourly_rate: number;
  /** 근무 달력에 등록된 이번 달 시간 (0이면 스케줄 없음) */
  scheduled_hours: number;
  profile: StaffProfile;
  /** "언제 일할 수 있어요?"의 답 — 달력에서 배정 가능한 사람을 뽑는 근거 */
  availability: AvailabilityWindow[];
  /** '월·수 09~14시' 같은 한 줄 요약 (서버가 만든다 — 화면·챗봇이 같은 문장을 쓰게) */
  availability_text: string;
  cost: LaborCost;
};

/** 달력의 근무 한 칸 */
export type Shift = {
  id: number;
  employee_id: number;
  name: string;
  role: string;
  date: string;
  start: string; // HH:MM
  end: string;   // HH:MM
  hours: number;
  /** 직원 대표 색 (없으면 화면이 팔레트에서 배정) */
  color: string | null;
  /** true=가능 시간 안 / false=가능 시간 밖 / null=가능 시간 미입력 */
  fits_availability: boolean | null;
};

export type CalendarDay = {
  date: string;
  weekday: number;      // 0=월
  weekday_label: string;
  shifts: Shift[];
  hours: number;
  available: {
    employee_id: number;
    name: string;
    role: string;
    color: string | null;
    windows: AvailabilityWindow[];
    already_assigned: boolean;
  }[];
};

export type StaffCalendar = {
  month: string;
  days: CalendarDay[];
  staff: {
    id: number;
    name: string;
    role: string;
    color: string | null;
    availability: AvailabilityWindow[];
    availability_text: string;
  }[];
  weekday_labels: string[];
};

export type StaffList = {
  staff: StaffMember[];
  month: string;
  total_gross: number;
  total_owner_burden: number;
  total_cost: number;
  total_hours: number;
  /** 근무시간을 아직 알 수 없는 직원 수 — 인건비가 0으로 보이는 이유를 설명할 때 쓴다 */
  unknown_hours_count: number;
  employment_types: { code: EmploymentType; label: string; note: string }[];
  insurance_types: { code: InsuranceType; label: string; note: string }[];
  min_wage: number;
};

export type WeeklyPayrollRow = {
  employee_id: number;
  name: string;
  role: string;
  hours: number;
  from_schedule: boolean;
  pay_type: 'hourly' | 'monthly';
  base_pay: number;
  weekly_holiday_pay: number;
  total: number;
};

export type WeeklyPayroll = {
  week_start: string;
  week_end: string;
  rows: WeeklyPayrollRow[];
  total: number;
  holiday_total: number;
  note: string;
};

/**
 * 등록 화면과 상세 화면이 함께 다루는 항목 — 두 화면이 같은 타입을 쓰게 해서
 * "추가할 땐 못 고르고 나중에 상세에서 또 고르는" 어긋남을 타입 단계에서 막는다.
 */
export type StaffEditable = {
  name: string;
  role: string;
  hourly_rate: number;
  /** 보낸 경우에만 통째로 교체된다 (빈 배열이면 가능 시간 전부 삭제) */
  availability: AvailabilityWindow[];
  /** 직원 대표 색 (#RRGGBB) — 달력의 점·선과 아바타를 같은 색으로 묶는다 */
  color: string;
  employment_type: EmploymentType;
  pay_type: 'hourly' | 'monthly';
  monthly_salary: number;
  weekly_hours: number;
  insurance: InsuranceType;
  weekly_holiday_pay: boolean;
  hired_on: string | null;
  memo: string | null;
};

/** 목록을 아직 못 불러왔을 때도 등록 화면이 같은 선택지를 보여줄 수 있게 두는 사본
 *  (문구·순서의 출처는 서버의 employment_types / insurance_types다) */
export const EMPLOYMENT_TYPE_FALLBACK: { code: EmploymentType; label: string; note: string }[] = [
  { code: 'part_time', label: '단시간 알바 (주 15시간 미만)', note: '주휴수당·4대보험 대상이 아니에요.' },
  { code: 'part_time_15', label: '알바 (주 15시간 이상)', note: '주휴수당이 발생하고 4대보험 가입 대상이에요.' },
  { code: 'full_time', label: '정규직 (주 40시간)', note: '주휴수당 포함, 4대보험 의무 가입입니다.' },
  { code: 'manager', label: '매니저 · 점장', note: '보통 월급제로 계약해요.' },
];

export const INSURANCE_TYPE_FALLBACK: { code: InsuranceType; label: string; note: string }[] = [
  { code: 'four', label: '4대보험', note: '사업주 부담이 임금의 약 10~11% 추가됩니다.' },
  { code: 'two', label: '2대보험 (고용·산재)', note: '단시간 근로자의 일반적인 형태예요.' },
  { code: 'none', label: '미가입 (3.3% 원천징수)', note: '사업소득 계약. 소급 징수될 수 있어요.' },
];

export const listStaff = (token: string) =>
  apiFetch<StaffList>('/api/v1/staff', { headers: auth(token) });

/** 직원 등록 + 고용 상세를 한 번에 — 응답은 목록 한 줄과 같은 모양이라 그대로 붙이면 된다 */
export const createStaff = (
  token: string,
  body: Partial<StaffEditable> & { name: string },
) =>
  apiFetch<StaffMember>('/api/v1/staff', {
    method: 'POST',
    headers: auth(token),
    body: JSON.stringify(body),
  });

/** 이름·시급·직책과 고용 상세를 함께 저장하고, 다시 계산된 그 직원 한 줄을 돌려받는다 */
export const saveStaffProfile = (
  token: string,
  employeeId: number,
  body: Partial<StaffEditable>,
) =>
  apiFetch<StaffMember>(`/api/v1/staff/${employeeId}/profile`, {
    method: 'PUT',
    headers: auth(token),
    body: JSON.stringify(body),
  });

/** 근무 가능 시간만 따로 저장 (통째 교체) */
export const saveAvailability = (
  token: string,
  employeeId: number,
  availability: AvailabilityWindow[],
) =>
  apiFetch<{ employee_id: number; availability: AvailabilityWindow[]; availability_text: string }>(
    `/api/v1/staff/${employeeId}/availability`,
    { method: 'PUT', headers: auth(token), body: JSON.stringify({ availability }) },
  );

/** 월 근무 달력 — 날짜별 배정 근무 + 그날 근무 가능한 직원 */
export const getStaffCalendar = (token: string, month?: string) =>
  apiFetch<StaffCalendar>(`/api/v1/staff/calendar${month ? `?month=${month}` : ''}`, {
    headers: auth(token),
  });

export const addShift = (
  token: string,
  body: { employee_id: number; date: string; start: string; end: string },
) =>
  apiFetch<Shift>('/api/v1/staff/shifts', {
    method: 'POST',
    headers: auth(token),
    body: JSON.stringify(body),
  });

/** 교대(담당 직원 교체)와 시간 변경 — employee_id를 보내면 그 근무를 다른 사람이 맡는다 */
export const updateShift = (
  token: string,
  shiftId: number,
  body: { employee_id?: number; date?: string; start?: string; end?: string },
) =>
  apiFetch<Shift>(`/api/v1/staff/shifts/${shiftId}`, {
    method: 'PUT',
    headers: auth(token),
    body: JSON.stringify(body),
  });

/** 근무 가능 시간 → 그 달 달력에 실제 근무로 채우기 (지난 날짜 제외, 중복은 건너뜀).
 *  같은 schedules 테이블을 쓰므로 '직원·스케줄'의 알바 근무 달력에도 그대로 나타난다. */
export const applyAvailability = (
  token: string,
  body: { month?: string; employee_id?: number } = {},
) =>
  apiFetch<{ month: string; created: number; skipped: number; shifts: Shift[] }>(
    '/api/v1/staff/shifts/apply-availability',
    { method: 'POST', headers: auth(token), body: JSON.stringify(body) },
  );

export const deleteShift = (token: string, shiftId: number) =>
  apiFetch<{ ok: boolean }>(`/api/v1/staff/shifts/${shiftId}`, {
    method: 'DELETE',
    headers: auth(token),
  });

/** 요일 라벨 — 0=월 … 6=일 (서버와 같은 순서) */
export const WEEKDAY_LABELS = ['월', '화', '수', '목', '금', '토', '일'];

export const getWeeklyPayroll = (token: string, weekStart?: string) =>
  apiFetch<WeeklyPayroll>(
    `/api/v1/staff/weekly-payroll${weekStart ? `?week_start=${weekStart}` : ''}`,
    { headers: auth(token) },
  );

// ---------------------------------------------------------------------------
// 월 급여·정산 배치 조회 — /operation/payroll/all의 빠른 대체
//
// 기존 경로는 서버가 직원 한 명마다 DB 왕복 5번을 반복해 직원 5명이면 5초가 걸렸고,
// 정산 계산은 내부에서 그걸 또 불러 '정산·급여' 카드가 10초 가까이 비어 있었다.
// /staff/monthly-payroll·month-settlement는 같은 숫자를 왕복 2~4번으로 만든다.
// 서버가 아직 옛 버전이면(404) 기존 경로로 폴백해 화면이 죽지 않게 한다.
// ---------------------------------------------------------------------------

/** /operation의 Payroll과 같은 모양 (프론트 화면 호환) */
export type MonthlyPayrollRow = {
  employee_id: number;
  employee_name: string;
  role: string;
  hourly_rate: number;
  period_start?: string;
  period_end?: string;
  total_work_hours: number;
  base_salary: number;
  weekly_holiday_allowance: number;
  estimated_salary: number;
  based_on_actual?: boolean;
};

export type MonthSettlement = {
  total_sales: number;
  total_expense: number;
  total_payroll: number;
  net_profit: number;
  year_month?: string;
  period_start?: string;
  period_end?: string;
  disclaimer?: string;
};

// 같은 화면의 여러 카드가 동시에 같은 달을 조회한다 — 진행 중인 요청은 하나로 합친다.
// (완료된 뒤에는 다시 부르면 새로 조회하므로 등록·수정 직후 갱신은 그대로 동작한다)
const inflight = new Map<string, Promise<unknown>>();
function shareInflight<T>(key: string, run: () => Promise<T>): Promise<T> {
  const hit = inflight.get(key) as Promise<T> | undefined;
  if (hit) return hit;
  const p = run().finally(() => inflight.delete(key));
  inflight.set(key, p);
  return p;
}

export const getMonthlyPayroll = (token: string, yearMonth?: string) =>
  shareInflight(`payroll:${yearMonth ?? ''}`, () =>
    apiFetch<MonthlyPayrollRow[]>(
      `/api/v1/staff/monthly-payroll${yearMonth ? `?year_month=${yearMonth}` : ''}`,
      { headers: auth(token) },
    ),
  );

/** 정산 + 직원별 급여를 요청 1번에 — '이번 달 정산·급여' 카드 전용 */
export const getMonthSettlement = (token: string, yearMonth?: string) =>
  shareInflight(`settlement:${yearMonth ?? ''}`, () =>
    apiFetch<{ settlement: MonthSettlement; payroll: MonthlyPayrollRow[] }>(
      `/api/v1/staff/month-settlement${yearMonth ? `?year_month=${yearMonth}` : ''}`,
      { headers: auth(token) },
    ),
  );

File: backend/app/api/v1/staff.py
"""직원 인건비 상세 API (백엔드 B)

기존 직원 등록(operation의 /employees)은 이름·시급·직책만 다룬다. 여기서는 그 위에
고용형태(주 15시간 미만/이상·정규·매니저), 급여형태(시급/월급), 보험(4대/2대/미가입)을
붙여 '실제로 매장에서 나가는 돈'을 계산한다.
"""

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.core.auth import get_current_user, require_owner
from app.models.user import User
from app.services.ai import staff_service

router = APIRouter(prefix="/staff", tags=["staff"])


class AvailabilityWindow(BaseModel):
    """근무 가능 시간 한 칸 — 요일(0=월 … 6=일)과 시작·종료 시각(정시 단위)"""

    day_of_week: int = Field(..., ge=0, le=6)
    start_hour: int = Field(9, ge=0, le=23)
    end_hour: int = Field(18, ge=1, le=24)


class ProfileUpdate(BaseModel):
    """직원 상세 화면에서 고칠 수 있는 모든 값 — 이름·시급(employees)과 고용 상세를 함께 받는다.

    등록 화면과 상세 화면이 같은 항목을 다뤄야 사장님이 "추가할 때 못 고른 걸 나중에
    또 찾아 들어가는" 일이 없다.
    """

    # 보낸 경우에만 통째로 교체된다 (빈 배열 = 가능 시간 전부 삭제)
    availability: Optional[List[AvailabilityWindow]] = None
    name: Optional[str] = None
    hourly_rate: Optional[int] = Field(None, ge=0)
    role: Optional[str] = None
    # 직원 대표 색 (#RRGGBB) — 근무 달력의 점·선과 아바타를 같은 색으로 맞추는 데 쓴다
    color: Optional[str] = Field(None, max_length=9)
    employment_type: Optional[str] = Field(None, description="part_time | part_time_15 | full_time | manager")
    pay_type: Optional[str] = Field(None, description="hourly | monthly")
    monthly_salary: Optional[int] = Field(None, ge=0)
    weekly_hours: Optional[float] = Field(None, ge=0, le=80)
    insurance: Optional[str] = Field(None, description="four | two | none")
    weekly_holiday_pay: Optional[bool] = None
    hired_on: Optional[str] = None
    memo: Optional[str] = None


class StaffCreate(ProfileUpdate):
    """직원 등록 — 이름만 필수, 나머지는 상세 화면과 같은 항목이며 생략하면 기본값."""

    name: str = Field(..., min_length=1)
    hourly_rate: int = Field(0, ge=0)


@router.get("", summary="직원 목록 + 고용 상세 + 월 인건비 추정")
def list_staff_api(
    month: Optional[str] = None,
    current_user: User = Depends(require_owner),
):
    """인건비는 근무 달력에 등록된 해당 월 근무시간 기준으로 계산된다.

    month는 YYYY-MM, 생략하면 이번 달.
    """
    return staff_service.list_staff(current_user.email, month=month)


@router.post("", summary="직원 등록 + 고용 상세 한 번에")
def create_staff_api(
    body: StaffCreate,
    current_user: User = Depends(require_owner),
):
    """등록과 상세 저장을 한 트랜잭션으로 묶는다 — 반쪽만 저장되는 직원이 안 생기게.

    반환값은 목록의 한 줄과 같은 모양(프로필 + 인건비 계산)이라 화면이 바로 붙일 수 있다.
    """
    try:
        return staff_service.create_staff(
            current_user.email, **body.model_dump(exclude_none=True),
        )
    except staff_service.StaffError as e:
        raise HTTPException(400, str(e))


@router.put("/{employee_id}/profile", summary="직원 기본 정보 + 고용 상세 저장")
def save_profile_api(
    employee_id: int,
    body: ProfileUpdate,
    current_user: User = Depends(require_owner),
):
    """저장된 결과를 목록 한 줄 모양(인건비 재계산 포함)으로 돌려준다."""
    try:
        return staff_service.save_profile(
            current_user.email, employee_id,
            **body.model_dump(exclude_none=True),
        )
    except staff_service.StaffError as e:
        raise HTTPException(400, str(e))


class AvailabilityUpdate(BaseModel):
    """근무 가능 시간만 따로 저장할 때 (통째 교체)"""

    availability: List[AvailabilityWindow] = Field(default_factory=list)


class ShiftCreate(BaseModel):
    employee_id: int
    date: str = Field(..., description="근무 날짜 YYYY-MM-DD")
    start: str = Field("09:00", description="시작 시각 HH:MM")
    end: str = Field("18:00", description="종료 시각 HH:MM (시작보다 이르면 익일로 본다)")


class ShiftUpdate(BaseModel):
    """교대·시간 변경 — 담당 직원을 바꾸는 것이 핵심이다"""

    employee_id: Optional[int] = Field(None, description="이 근무를 대신할 직원 ID (교대)")
    date: Optional[str] = None
    start: Optional[str] = None
    end: Optional[str] = None


@router.put("/{employee_id}/availability", summary="직원 근무 가능 시간 저장 (통째 교체)")
def save_availability_api(
    employee_id: int,
    body: AvailabilityUpdate,
    current_user: User = Depends(require_owner),
):
    try:
        windows = staff_service.save_availability(
            current_user.email, employee_id,
            [w.model_dump() for w in body.availability],
        )
        return {"employee_id": employee_id, "availability": windows,
                "availability_text": staff_service.describe_windows(windows)}
    except staff_service.StaffError as e:
        raise HTTPException(400, str(e))


@router.get("/calendar", summary="월 근무 달력 — 날짜별 배정 근무 + 그날 가능한 직원")
def calendar_api(
    month: Optional[str] = None,
    current_user: User = Depends(require_owner),
):
    """month는 YYYY-MM, 생략하면 이번 달."""
    try:
        return staff_service.month_calendar(current_user.email, month=month)
    except staff_service.StaffError as e:
        raise HTTPException(400, str(e))


@router.post("/shifts", summary="근무 배정 추가")
def add_shift_api(
    body: ShiftCreate,
    current_user: User = Depends(require_owner),
):
    try:
        return staff_service.add_shift(
            current_user.email, body.employee_id, body.date, body.start, body.end,
        )
    except staff_service.StaffError as e:
        raise HTTPException(400, str(e))


class ApplyAvailability(BaseModel):
    """가능 시간 → 달력 반영 (해당 월, 지난 날짜 제외, 중복은 건너뜀)"""

    month: Optional[str] = Field(None, description="YYYY-MM, 생략하면 이번 달")
    employee_id: Optional[int] = Field(None, description="비우면 매장 전체 직원")


@router.post("/shifts/apply-availability", summary="근무 가능 시간대로 달력 채우기")
def apply_availability_api(
    body: ApplyAvailability,
    current_user: User = Depends(require_owner),
):
    """등록해 둔 가능 시간을 그 달의 실제 근무로 만든다 — 기존 근무 달력에 그대로 뜬다."""
    try:
        return staff_service.apply_availability(
            current_user.email, month=body.month, employee_id=body.employee_id,
        )
    except staff_service.StaffError as e:
        raise HTTPException(400, str(e))


@router.put("/shifts/{schedule_id}", summary="근무 교대 · 시간 변경")
def update_shift_api(
    schedule_id: int,
    body: ShiftUpdate,
    current_user: User = Depends(require_owner),
):
    """employee_id를 보내면 그 근무를 다른 직원이 대신하는 것으로 바뀐다 (출퇴근 기록 유지)."""
    try:
        return staff_service.update_shift(
            current_user.email, schedule_id,
            employee_id=body.employee_id, day=body.date, start=body.start, end=body.end,
        )
    except staff_service.StaffError as e:
        raise HTTPException(400, str(e))


@router.delete("/shifts/{schedule_id}", summary="근무 배정 삭제")
def delete_shift_api(
    schedule_id: int,
    current_user: User = Depends(require_owner),
):
    try:
        staff_service.remove_shift(current_user.email, schedule_id)
        return {"ok": True}
    except staff_service.StaffError as e:
        raise HTTPException(400, str(e))


@router.get("/monthly-payroll", summary="전 직원 월 예상 급여 (배치 계산 — 직원 수와 무관하게 왕복 2번)")
def monthly_payroll_api(
    year_month: Optional[str] = None,
    current_user: User = Depends(require_owner),
):
    """/operation/payroll/all과 같은 모양·같은 기준. 그 경로는 직원마다 DB 왕복 5번이 붙어
    직원 5명이면 5초가 걸렸다. year_month는 YYYY-MM, 생략하면 이번 달."""
    try:
        return staff_service.monthly_payroll(current_user.email, year_month=year_month)
    except staff_service.StaffError as e:
        raise HTTPException(400, str(e))


@router.get("/month-settlement", summary="월 손익 정산 + 직원별 급여 한 번에 (배치 계산)")
def month_settlement_api(
    year_month: Optional[str] = None,
    current_user: User = Depends(require_owner),
):
    """'정산·급여' 카드가 요청 1번(DB 왕복 4번)으로 그려지게 정산과 급여를 함께 준다.
    집계 기준은 기존 /operation/settlements/calculate와 동일하다."""
    try:
        return staff_service.month_settlement(current_user.email, year_month=year_month)
    except staff_service.StaffError as e:
        raise HTTPException(400, str(e))


@router.get("/weekly-payroll", summary="이번 주 주급 정산 (스케줄 기준)")
def weekly_payroll_api(
    week_start: Optional[str] = None,
    current_user: User = Depends(require_owner),
):
    try:
        return staff_service.weekly_payroll(current_user.email, week_start=week_start)
    except ValueError as e:
        raise HTTPException(400, str(e))
